backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,318 @@
+                year_as_int = current_year if sub_link_of_each_year == "current_year_fire_data" \
+                                         else int(sub_link_of_each_year.split("_")[0])
+                # ignore years that before the start year
+                # filter out links that is too old
+                if year_as_int >= start_year:
+                    for state in self.states:
+                        # for each state, extract fire events in that year at that state
+                        fires.extend(self._extract_fire_events_in_state(year_as_int, state))
+            self._filter_out_invalid_fire_events(fires)
+            return fires
+
+    @staticmethod
+    def _download_fire_record(single_file_name: str, used_folder_names: Set[str], url_of_fire_event: str):
+        """
+        Downloads a single fire record from the website and puts its sub-files into a folder.
+        :param single_file_name: e.g. "ca_cedar_20170810_0000_dd83.CPG"
+        :param used_folder_names:e.g. ("ca_cedar_20170810_0000_dd83", "ca_cedar_20170812_2145_dd83")
+        :param url_of_fire_event:e.g "https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/2017_fire_data/California/Cedar/"
+        """
+        if not ".zip" == single_file_name[-4:]:
+            # if it is a zip don't download
+            folder_name = single_file_name.split(".")[0]
+            if folder_name not in used_folder_names:
+                # if the folder name is not in used_folder_names, which means it is a new folder
+                # then create the folder and add it into the used_folder_names set, to avoid duplicate folders
+                used_folder_names.add(folder_name)
+                os.makedirs(os.path.join(FIRE_DATA_DIR, folder_name))
+            out_path = os.path.join(FIRE_DATA_DIR, folder_name)
+            FireCrawler._download_single_file(url_of_fire_event, single_file_name, out_path)
+
+    @staticmethod
+    @retry(retry_on_exception=requests.exceptions.RequestException, stop_max_attempt_number=3)
+    def _extract_file_names_in_fire_event(url_of_fire_event: str) -> List[str]:
+        """
+        Extracts all file names from a fire_event page. Return a list of all file names.
+        :param url_of_fire_event: the url of the fire_event
+                e.g. https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/2017_fire_data/California/Antelope/
+        :return: e.g. ["ca_antelope_20170706_1930_dd83.CPG", "ca_antelope_20170706_1930_dd83.dbf"]
+        """
+        fire_page = FireCrawler._get_url(url_of_fire_event, url_of_fire_event, "crawl()")
+        return FireCrawler.RE_EXTRACT_FILES_IN_FIRE_EVENTS.findall(fire_page)
+
+    def crawl(self, url_of_fire_event: str):
+        """
+        Takes the url string of one fire and start to crawl all files.
+        :param url_of_fire_event: str
+                e.g. "https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/2016_fire_data/California/Happy_Camp"
+        """
+        used_folder_names = set()
```

Comment:
this set is initialized to be empty, to used to check for used folder names.  but there could be existing folders when the crawler is first initialized right? don't you need to do a file system scan to get those existing ones?